pkg/services/ngalert/schedule/schedule.go

```diff
@@ -431,18 +445,25 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 	for {
 		select {
 		// used by external services (API) to notify that rule is updated.
-		case lastVersion := <-updateCh:
+		case ctx := <-updateCh:
+			// we can continue directly if the rule is paused because in the next tick it will be picked up. The state
+			// will be cleaned up and the routine will be stopped and removed from the registry
+			if ctx.IsPaused {
```

Comment:
the sequence of the command is still not correct. I think the rule version check should be the first. It's a guard for concurrent access to the rule routine. 
So, it should always update the version of the rule to the latest or ignore if the version is less that the current.
Imagine a hypothetical situation: update comes via API route, version Y, which makes the rule to be paused. The scheduler is already in the middle of processing of tick and has read the rules from db, so it process a rule of version X. 
Therefore, in the current state, the rule routine will reset the state but not change the version and millisecond later the scheduler will send an evaluation for rule of version X and the state will be populated again. Then the next tick, scheduler process version Y and stops the rule routine, which again resets the state.

Reply:
apparently eval path of the routine has a bug! and it will evaluate a rule even if the local version is greater than the one that was received. We can fix it in a follow-up PR.